source/extensions/filters/http/ext_authz/ext_authz.cc

```diff
@@ -65,14 +82,36 @@ void Filter::initiateCall(const Http::HeaderMap& headers) {
   initiating_call_ = false;
 }
 
-Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool) {
+Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool end_stream) {
   request_headers_ = &headers;
+  const auto* method = headers.Method();
+  buffer_data_ = config_->withRequestBody() &&
+                 !(end_stream || (method && isHeaderOnlyMethod(method->value().getStringView())));
+
+  if (buffer_data_) {
+    ENVOY_STREAM_LOG(debug, "ext_authz is buffering", *callbacks_);
+    if (!config_->allowPartialMessage()) {
+      callbacks_->setDecoderBufferLimit(config_->maxRequestBytes());
+    }
+    return Http::FilterHeadersStatus::StopIteration;
+  }
+
   initiateCall(headers);
   return filter_return_ == FilterReturn::StopDecoding ? Http::FilterHeadersStatus::StopIteration
                                                       : Http::FilterHeadersStatus::Continue;
 }
 
-Http::FilterDataStatus Filter::decodeData(Buffer::Instance&, bool) {
+Http::FilterDataStatus Filter::decodeData(Buffer::Instance&, bool end_stream) {
+  if (buffer_data_) {
+    if (!end_stream && !isBufferFull()) {
+      return Http::FilterDataStatus::StopIterationAndBuffer;
+    }
+
+    // Call the authorization service.
+    ENVOY_STREAM_LOG(debug, "ext_authz finished buffering", *callbacks_);
+    initiateCall(*request_headers_);
```

Comment:
If the request is small and with trailer (i.e. end_stream is false), then `initiateCall` won't be called? Add a test case for that.

Reply:
👍 Great question! Thanks